### PR TITLE
feat: add gMLP Implementation

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
Adds gMLP Implementation (#12)

### CODE REVIEW

File: src/jax/modules/gmlp.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        """Setup the SpatialGatingUnit layer based on the input shape"""
+        self.dense = nn.Dense(features=self.dim)
+
+    @nn.compact
+    def __call__(self, inputs: ArrayLike, *args, **kwargs) -> ArrayLike:
+        """Compute a forward pass through the SpatialGatingUnit"""
+        # pylint: disable=invalid-name
+        u, v = inputs.split(2, axis=-1)
+        v = nn.LayerNorm()(v)
+        v = self.dense(v.transpose((0, 2, 1))).transpose((0, 2, 1))
```
Comment (soumik12345) [inline src/jax/modules/gmlp.py:34]: **Nitpick:** the spatial projection could be written in a little more verbose manner.

File: src/pytorch/modules/gmlp.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Forward pass for SpatialGatingUnit"""
+        # pylint: disable=invalid-name
+
+        u, v = inputs.chunk(chunks=2, dim=-1)
+
+        v = nn.LayerNorm(normalized_shape=v.shape, eps=1e-6)(v)
+        v = v.permute(0, 2, 1)
+        v = nn.Linear(in_features=self.dim, out_features=self.dim, bias=True)(v)
+        v = v.permute(0, 2, 1)
```
Comment (soumik12345) [inline src/pytorch/modules/gmlp.py:31]: **Nitpick:** add a comment mentioning that this is spatial projection, also, make the flax module similarly verbose.

File: src/pytorch/modules/gmlp.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        - https://arxiv.org/abs/2105.08050v2
+    """
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Forward pass for gMLPBlock"""
+        shortcut = inputs
+        norm = nn.LayerNorm(normalized_shape=inputs.shape, eps=1e-6)(inputs)
+        in_proj = nn.Linear(
+            in_features=inputs.shape[-1], out_features=inputs.shape[-1] * 2, bias=True
+        )(norm)
```
Comment (soumik12345) [inline src/pytorch/modules/gmlp.py:50]: **Nitpick:** add a comment mentioning that this is channel projection.

File: src/tensorflow/modules/gmlp.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+        # Attributes
+        self.dim = dim
+
+    def call(self, inputs: tf.Tensor, *args, **kwargs) -> tf.Tensor:
+        """Forward pass for SpatialGatingUnit"""
+        # pylint: disable=invalid-name
+        u, v = tf.split(inputs, num_or_size_splits=2, axis=-1)
+        v = tf.keras.layers.LayerNormalization(epsilon=1e-6)(v)
+        v = tf.linalg.matrix_transpose(v)
```
Comment (soumik12345) [inline src/tensorflow/modules/gmlp.py:26]: Why not simply use [`tf.transpose`](https://www.tensorflow.org/api_docs/python/tf/transpose)?

File: src/tensorflow/modules/gmlp.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        v = tf.keras.layers.Dense(self.dim, bias_initializer="ones")(v)
```
Comment (soumik12345) [inline src/tensorflow/modules/gmlp.py:27]: **Nitpick:** add a comment mentioning these lines as spatial projection.

File: src/tensorflow/modules/gmlp.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        v = tf.linalg.matrix_transpose(v)
```
Comment (soumik12345) [inline src/tensorflow/modules/gmlp.py:28]: Why not simply use [`tf.transpose`](https://www.tensorflow.org/api_docs/python/tf/transpose)?

**Review by soumik12345 — CHANGES_REQUESTED** [review]
Hi @SauravMaheshkar, suggested a few changes...

File: src/tensorflow/modules/gmlp.py (hunk 4 above)
Comment (SauravMaheshkar) [reply src/tensorflow/modules/gmlp.py:26]: Thanks for this catch, will fix.

File: src/jax/modules/gmlp.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    @nn.compact
+    def __call__(self, inputs: ArrayLike, *args, **kwargs) -> ArrayLike:
+        """Compute a forward pass through the SpatialGatingUnit"""
+        # pylint: disable=invalid-name
+        u, v = inputs.split(2, axis=-1)
+        v = nn.LayerNorm()(v)
+        v = v.transpose((0, 2, 1))
+        v = self.dense(v)
+        v = v.transpose((0, 2, 1))
```
Comment (soumik12345) [inline src/jax/modules/gmlp.py:36]: ❤️

**Review by soumik12345 — APPROVED** [review]
Goot to go 🚀

**Pull request merged**